lib/api/controllers/funnelController.js

```diff
@@ -80,51 +86,65 @@ function FunnelController (kuzzle) {
-        .then(response => callback(null, response.responseObject))
+      processRequest(kuzzle, this, request)
+        .then(processResult => callback(null, processResult))
         .catch(err => {
           kuzzle.pluginsManager.trigger('log:error', err);
-          callback(err, new ResponseObject(requestObject, err));
+
+          request.setError(err);
+          callback(err, request);
 
           this.handleErrorDump(err);
 
           return null;
         })
         .finally(() => {
           this.concurrentRequests--;
-          delete this.pendingRequests[requestObject.requestId];
+          delete this.pendingRequests[request.id];
         });
 
       return null;
     }
 
     /*
-      If kuzzle is overloaded, check the requests cache.
-      There are two possibilities:
-        1- the cache limit has not been reached. A warning is emitted, and the
-           request is cached and will be played as soon as the maxConcurrentRequests
-           property allows it
-
-        2- the number of cached requests is equal to the maxRetainedRequests property.
-           The request is then discarded and an error is returned to the sender
+     If kuzzle is overloaded, check the requests cache.
+     There are two possibilities:
+
+     1- the cache limit has not been reached: the request is cached
+     and will be played as soon as the maxConcurrentRequests
+     property allows it
+
+     2- the number of cached requests is equal to the maxRetainedRequests property.
+     The request is then discarded and an error is returned to the sender
      */
-    if (!this.overloaded) {
-      this.overloaded = true;
-      setTimeout(() => playCachedRequests(kuzzle, this), 10);
-    }
+    if (this.cachedItems >= kuzzle.config.server.maxRetainedRequests) {
+      const error = new ServiceUnavailableError('Request discarded: Kuzzle Server is temporarily overloaded');
 
-    if (this.cachedRequests >= kuzzle.config.server.maxRetainedRequests) {
-      error = new ServiceUnavailableError('Request discarded: Kuzzle Server is temporarily overloaded');
       kuzzle.pluginsManager.trigger('log:error', error);
-      callback(error, new ResponseObject(requestObject, error));
-
+      request.setError(error);
+      callback(error, request);
       return null;
     }
 
-    this.cachedRequests++;
-    this.requestsCache.push({requestObject, userContext, callback});
+    this.cachedItems++;
+    this.requestsCache.push(new CacheItem(request, callback));
+
+    if (!this.overloaded) {
+      this.overloaded = true;
+
+      /*
+       /!\ Invoking this function here with setTimeout() leads to V8 deoptimizing
+       the entire funnelExecute method (as of node.js 6.9.1),
+       because of an "out of bounds" heuristic error (caused by node's
+       setTimeout code written in JS? this needs further investigations)
+
+       We get better performances executing this function once for nothing each
```

Comment:
That's because I've mixed up the beginning and the end of this sentence when I wrote it. I fixed it.